c2rust-analyze/src/rewrite/apply.rs

```diff
@@ -0,0 +1,506 @@
+            // at least partially overlap the current item.  These should normally be ancestors of
+            // the current item on the stack, but it's also possible that the current item
+            // erroneously partially overlaps a previous item.
+            while stack.last().map_or(false, |rt| rt.span.hi() <= lo) {
+                let rt = stack.pop().unwrap();
+                commit(&mut stack, &mut out, rt);
+            }
+
+            // Check for error cases.
+            let siblings = if let Some(parent) = stack.last() {
+                debug_assert!(parent.span.overlaps(span));
+                if !parent.span.contains(span) {
+                    // `span` overlaps `parent.span`, but isn't contained within it.
+                    errs.push((span, rw, RewriteError::PartialOverlap(parent.span)));
+                    continue;
+                }
+                &parent.children
+            } else {
+                &out
+            };
+
+            if let Some(prev) = siblings.last() {
+                debug_assert!(prev.span.lo() <= span.lo());
+                if prev.span.overlaps(span) {
+                    errs.push((span, rw, RewriteError::PartialOverlap(prev.span)));
+                    continue;
+                }
+            }
+
+            // TODO: check that all children are emitted somewhere in the `Rewrite`
+            // (e.g. given the rewrite `f(x + y) -> x`, we can't also have a rewrite on `x + y`,
+            // because that expression is discarded as part of the outer rewrite)
+            //
+            // Specifically: if there is a parent, check that either the current item's span is
+            // contained in the parent span and the parent's rewrite contains `Rewrite::Identity`,
+            // or the current item's span is contained in the span of some `Rewrite::Sub` in
+            // the parent rewrite.  If this doesn't hold, then produce `RewriteError::Discarded`.
+
+            // Push a new node onto the stack.
+            stack.push(RewriteTree {
+                span,
+                rw,
+                children: Vec::new(),
+            });
+        }
+
+        while let Some(rt) = stack.pop() {
+            commit(&mut stack, &mut out, rt);
+        }
+
+        (out, errs)
+    }
+}
+
+/// Split `rts` into the portion before `span`, the portion overlapping `span`, and the portion
+/// after `span`.  Nodes within `rts` should not overlap each other, and the list must be sorted by
+/// span; otherwise, the result are unspecified.
+fn partition_nodes<'a>(
+    rts: &'a [RewriteTree],
+    span: Span,
+) -> (&'a [RewriteTree], &'a [RewriteTree], &'a [RewriteTree]) {
+    let lo = span.lo();
+    let hi = span.hi();
+
+    // Collect nodes from the front of `rts` until we find one that overlaps or comes after `span`.
+    let i = rts
+        .iter()
+        .position(|rt| rt.span.hi() > lo)
```

Comment:
intuitively i'd expect the first overlapping span to have the property `rt.span.lo() <= hi`. what am i missing?